flow/layers/layer.h

```diff
@@ -171,7 +175,14 @@ class Layer {
     paint_bounds_ = paint_bounds;
   }
 
```

Comment:
Let's take this opportunity to add some documentations to `Layer::paint_bounds`. Before retained layers are introduced, `Layer::paint_bounds` is simple as each layer only exists for one frame with a single context. After retained layers, we need to explicitly document that `Layer::paint_bounds` must be independent of its parent context as the layer subtree can be retained and used in another context.

As it's context-independent, I wonder if `Preroll` is the best place to compute `paint_bounds` since there's a`PrerollContext`, and one might be tempted to use it. Shall we compute the paint bounds during the layer construction phase? For example, during `SceneBuilder` phase, when all children layers are added, call a method such as `Layer::finalize` to notify that no more children are expected to be added, and the `paint_bounds` can now be computed in a context-independent way.

Reply:
We've been discussing these issues over on the issue page linked above.

I think we are pretty much good to go on pre-computing the bounds at scene construction time, but there is still the issue of the shadow bounds depending on device pixel ratio which I think might be a bug. I briefly played with pre-computing the bounds as a variation of this PR, but decided we can do that in a separate operation - the main objective here is to get rid of the pre-mature culling while not breaking anything else.

I'll add some comments to paint_bounds().

Reply:
I updated the comments in the recent commit.